### PR TITLE
Spread call/stakedao

### PR DESCRIPTION
**Pull request by ugolino**

### CODE REVIEW

File: contracts/actions/ShortOTokenActionWithSwap.sol [hunk 1]
```diff
@@ -268,6 +268,10 @@ contract ShortOTokenActionWithSwap is IAction, AirswapBase, RollOverBase {
             referralCode
         );
 
+    // TODO : do this before repaying flash loan to avoid multiple fee
+    // convert the weth received as premium to sdeCRV
+    _wethToSdEcrv();
+
```
Comment (aparnakr) [inline contracts/actions/ShortOTokenActionWithSwap.sol:274]: can you explain why you do this here and not just in the executeOperation? :)

File: test/mainnet-fork/short-vault-loss.ts [hunk 2]
```diff
@@ -126,7 +127,7 @@ describe('Mainnet Fork Tests', function() {
 
     depositor1 = _depositor1;
     depositor2 = _depositor2;
-    depositor3 = _depositor3;
+    depositor3 = _depositor3
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:130]:
```suggestion
    depositor3 = _depositor3;
```

File: test/mainnet-fork/short-vault-loss.ts [hunk 3]
```diff
@@ -404,73 +440,124 @@ describe('Mainnet Fork Tests', function() {
...
+
       const premiumInSdecrv = premium.mul(95).div(100);
-      const expectedTotal = vaultSdecrvBalanceBefore.add(premiumInSdecrv);
+      // estimating fee for flash loan and wrapping ~10%
+      const netPremiumInSdecrv = premium.mul(90).div(100)
+      const expectedTotal = vaultSdecrvBalanceBefore.add(netPremiumInSdecrv);
       expectedSdecrvBalanceInAction = expectedSdecrvBalanceInVault.add(premiumInSdecrv);
-      const sellAmount = (collateralAmount.div(10000000000)).toString(); 
-      optionsSold =  collateralAmount.div(10000000000);
+      // const sellAmount = (collateralAmount.div(10000000000)).toString();
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:453]:
```suggestion
```

File: test/mainnet-fork/short-vault-loss.ts [hunk 4]
```diff
@@ -404,73 +440,124 @@ describe('Mainnet Fork Tests', function() {
...
+      
+      const longStrikePrice = await longOtoken.strikePrice();
+      const shortStrikePrice = await shortOtoken.strikePrice();
+      
+      // // ((((longStrike).sub(shortStrike)).mul(1e10)).div(longStrike))
+      const collateralRequiredPerOption = (longStrikePrice.sub(shortStrikePrice).mul(1e10).div(longStrikePrice));
 
-      const marginPoolBalanceOfStakeDaoLPBefore = await stakeDaoLP.balanceOf(marginPoolAddess);
+      const sdcrvAmount = collateralAmount;
+      // const sellAmount = (collateralAmount.add(collateralAmount)).div(1e10).toString();
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:462]:
```suggestion
```

File: test/mainnet-fork/short-vault-loss.ts [hunk 5]
```diff
@@ -404,73 +440,124 @@ describe('Mainnet Fork Tests', function() {
...
+      expect( (actionVault.shortOtokens[0]), 'Action does not have the correct short otoken' ).to.be.equal(shortOtoken.address);
+      expect( (actionVault.longOtokens[0]), 'Action does not have the correct long otoken' ).to.be.equal(longOtoken.address);
+      expect( (actionVault.longAmounts[0]), 'Action does not have the correct amount for long otoken' ).to.be.equal(sellAmount);
+      expect( (actionVault.shortAmounts[0]), 'Action does not have the correct amount for short otoken' ).to.be.equal(sellAmount);
+      expect( (actionVault.collateralAssets[0]), 'Action does not have the right collateral' ).to.be.equal(stakeDaoLP.address);
+      expect( (actionVault.collateralAmounts[0]), 'Action does not have the correct amount of required collateral' ).to.be.lte( collateralAmount );
+
+
+      // check the otoken balance of the MM
+      // expect( (await longOtoken.balanceOf(action1.address)), 'Mismatch of longOtokens' ).to.be.equal(sellAmount);
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:532]:
```suggestion
```

File: test/mainnet-fork/short-vault-loss.ts [hunk 6]
```diff
@@ -404,73 +440,124 @@ describe('Mainnet Fork Tests', function() {
...
+      // expect(await shortOtoken.balanceOf(counterpartyWallet.address), 'incorrect otoken balance sent to counterparty').to.be.equal(
+      //   sellAmount
+      // );
 
-      const marginPoolBalanceOfStakeDaoLPAfter = await stakeDaoLP.balanceOf(marginPoolAddess);
+      const marginPoolBalanceOfStakeDaoLPAfter = await stakeDaoLP.balanceOf(marginPoolAddress);
 
-      // check sdecrv balance in opyn 
-      expect(marginPoolBalanceOfStakeDaoLPAfter, 'incorrect balance in Opyn').to.be.equal(marginPoolBalanceOfStakeDaoLPBefore.add(collateralAmount));
+      // // check sdecrv balance in opyn
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:542]:
```suggestion
      // check sdecrv balance in opyn 
```

File: test/mainnet-fork/short-vault-loss.ts [hunk 7]
```diff
@@ -672,24 +784,37 @@ describe('Mainnet Fork Tests', function() {
...
+
+        const longStrikePrice = await longOtoken.strikePrice();
+        const shortStrikePrice = await shortOtoken.strikePrice();
+
+        const payout =  sellAmount.mul(sdecrvToETHPrice)
+                                        .mul(1e10)
+                                        .div(sdecrvToETHPrice)
+                                        .mul( shortStrikePrice )
+                                        .div( sdecrvToETHPrice )
+
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:803]:
```suggestion
        const payout =  sellAmount.mul(1e10)
                                        .mul( shortStrikePrice )
                                        .div( sdecrvToETHPrice )

```

Comment (aparnakr) [reply test/mainnet-fork/short-vault-loss.ts:803]: I don't remember the exact formula, but might be worth checking that from gamma and making sure it matches that :D

**Review by aparnakr — APPROVED** [review]

**Comment by aparnakr** [comment]
Great job on this @ugolino! :D Left some comments to help you! This is a really challenging vault to build!

File: test/mainnet-fork/short-vault-loss.ts [hunk 8]
```diff
@@ -301,35 +309,60 @@ describe('Mainnet Fork Tests', function() {
...
+        const shortOtokenAddress = await otokenFactory.getOtoken(
+          weth.address,
+          usdc.address,
+          stakeDaoLP.address,
+          shortOtokenStrikePrice,
+          expiry,
+          false
+        );
+
+        shortOtoken = (await ethers.getContractAt(
```
[the remark this answers is not in the document]
Comment (ugolino) [reply test/mainnet-fork/short-vault-loss.ts:335]: yeah. agree!

File: test/mainnet-fork/short-vault-loss.ts [hunk 9]
```diff
@@ -390,11 +425,12 @@ describe('Mainnet Fork Tests', function() {
 
     it('owner commits to the option', async () => {
       expect(await action1.state()).to.be.equal(ActionState.Idle);
-      await action1.commitOToken(otoken.address);
+      await action1.commitSpread(shortOtoken.address, longOtoken.address);
```
[the remark this answers is not in the document]
Comment (ugolino) [reply test/mainnet-fork/short-vault-loss.ts:428]: @KonGol33 should have worked on this

**Pull request merged**